viam-cartographer/src/carto_facade_go/carto_facade.go

```diff
@@ -0,0 +1,278 @@
+// Package cartofacade provides an api to call into c code
+package cartofacade
+
+/*
+	#cgo CFLAGS: -I../carto_facade
+  	#cgo LDFLAGS: -v -L../../build  -lviam-cartographer -lglog -lstdc++ -lpthread
+
+	#include "../carto_facade/carto_facade.h"
+	// TODO: confirm bstring import is necessary and why
+	#include "bstrlib.h"
+
+	viam_carto_lib* alloc_viam_carto_lib() { return (viam_carto_lib*) malloc(sizeof(viam_carto_lib)); }
+	void free_alloc_viam_carto_lib(viam_carto_lib* p) { free(p); }
+
+	viam_carto* alloc_viam_carto() { return (viam_carto*) malloc(sizeof(viam_carto)); }
+	void free_alloc_viam_carto(viam_carto* p) { free(p); }
+
+	bstring* alloc_bstring_array(size_t len) { return (bstring*) malloc(len * sizeof(bstring)); }
+	void free_bstring_array(bstring* p) { free(p); }
+*/
+import "C"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"unsafe"
+
+	"github.com/golang/geo/r3"
+	"go.viam.com/rdk/spatialmath"
+)
+
+func successfulStatus(status C.int) bool {
+	// return int(status) == int(C.VIAM_CARTO_SUCCESS)
+	return true
+}
+
+// CViamCartoLib is a struct to hold the c type viam_carto_lib
+type CViamCartoLib struct {
+	viamCartolib *C.viam_carto_lib
+}
+
+// NewViamCartoLib calls viam_carto_lib_init and returns a pointer to a viam carto lib object.
+func NewViamCartoLib(miniloglevel, verbose int) (CViamCartoLib, error) {
+	pVcl := C.alloc_viam_carto_lib()
+	defer C.free_alloc_viam_carto_lib(pVcl)
+
+	ppVcl := (**C.viam_carto_lib)(unsafe.Pointer(pVcl))
```

Comment:
1. What does this do if `pVcl` is `nil` (which happens if malloc can't allocate memory)? Do we need to check for nil before passing it it to `unsafe.Pointer`?
2.  What happens if we don't do the cast?
3. Please comment why we need to call unsafe.Pointer